### PR TITLE
Refactor parser to parse JSON to DTOs instead of Impls, Allow adding rules with externally provided UID, Remove requirement for Manifest header "Automation-ResourceType"

### PR DESCRIPTION
**Pull request by adimova**

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/AutomationFactory.java [hunk 1]
```diff
@@ -110,4 +136,18 @@ public Rule createRule(List<Trigger> triggers, //
...
+     * @param uid is an unique identifier of the {@link Rule} instance.
+     * @param ruleTemplateUID is an unique identifier of the {@link RuleTemplate} instance.
+     * @param configurations are configuration values of the {@link Rule} instance.
+     * @return {@link Rule} instance.
+     * @throws Exception is thrown when the configuration values for required configuration properties are missing or
+     *             their type is wrong.
+     */
+    public Rule createRule(String uid, String ruleTemplateUID, Map<String, Object> configurations);
+
+    public Rule createRule(RuleDTO ruleDTO);
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/AutomationFactory.java:151]: Please add some JavaDoc here as well

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Converter.java [hunk 2]
```diff
@@ -9,14 +9,21 @@
 
 import java.util.Dictionary;
 
+/**
+ * This interfase provides opportunity to use different parsers for converting JsonObject to Dictionary.
+ *
+ * @author Vasil Ilchev - Initial Contribution
+ *
+ */
 public interface Converter {
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Converter.java:18]: Do we at all need this separate interface? Couldn't the method be moved into the Parser interface?

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Parser.java [hunk 3]
```diff
@@ -17,7 +17,7 @@
  *
  * @author Ana Dimova - Initial Contribution
  */
-public interface Parser {
+public interface Parser<T> {
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Parser.java]: hm, doesn't this mean that we need different Parser instances for rules, module types, etc?
I so far assumed that we register Parsers as services that can parse all of these things.

File: bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Converter.java (hunk 2 above)
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Converter.java:18]: They are different things. The parser parses JSON to DTO objects, the Converter parses string value of context parameter of ConfigDescriptionParameter when resolve connection between custom and system module types.
It is a  service which parses JSON to Map which is used by the AbstractMosuleHandler.

Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Converter.java:18]: So the Converter is yet another service that parses JSON, which is there for just a small very special use case.
Actually, I cannot remember that we ever discussed having JSON within config description parameters. Looking at your sample files, this imho looks rather ugly. If you now thing of other formats (e.g. XML) where you would put such JSON into feels pretty weird.
I'd suggest to discuss whether we really need it that way or if we could come up with a special syntax that suits our needs and which could be processed within the AbstractModuleHandler directly (without requiring any external services).

Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/parser/Converter.java:18]: I have added https://github.com/marinmitev/smarthome/issues/55 for this discussion.

File: bundles/automation/org.eclipse.smarthome.automation.parser.json/src/main/java/org/eclipse/smarthome/automation/parser/json/ModuleJSONParser.java [hunk 4]
```diff
@@ -38,15 +39,12 @@
...
-     * @param moduleTypesStatus
-     * @param triggers is a List which has to be filled with {@link Trigger} objects created from json.
+     * @param triggerModules is a List which has to be filled with {@link Trigger} objects created from json.
      * @param onSection is a {@link JSONArray} object containing {@link JSONObject}s representing {@link Trigger}s in
      *            json format.
      * @return
      */
-    static boolean createTrigerModules(Status status, AutomationFactory automationFactory,
-            LinkedHashSet<Status> moduleTypesStatus, List<Trigger> triggers, JSONArray onSection) {
+    static boolean createTrigerModules(Status status, List<TriggerDTO> triggerModules, JSONArray onSection) {
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.parser.json/src/main/java/org/eclipse/smarthome/automation/parser/json/ModuleJSONParser.java:47]: typo: createTriggerModules

**Pull request merged**